### ISSUE TITLE
Assert in GC_mark_from may fail erroneously

### ISSUE THREAD

**Issue by GaborDrescher** [ivmai/bdwgc#124]
Hi,
there may be a possible bug here: https://github.com/ivmai/bdwgc/blob/8ffc3db1c56c4ef327538fe8a667a2abebbce773/mark.c#L638-L639 in `GC_mark_from`, the assertion:
```
GC_ASSERT(descr < (word)GC_greatest_plausible_heap_addr
- (word)GC_least_plausible_heap_addr);
```
is only valid for regions allocated from the GC heap, but `GC_mark_from` iterates over all regions pushed on the mark stack, including data, bss, stack regions and custom regions. If one of these regions exceeds the GC heap size, this assertion may fail erroneously.

Background: We ported bdwgc to run on our research platform that uses a kind of software-controlled cache. This cache may contain pointers to GC-managed objects, we `GC_push_all` this region so the GC can scan it for pointers. In some configurations this cache is larger that the GC-managed memory and in these cases we trigger this assertion.

**Comment by ivmai** [comment]
I guess the expression should be changed to: descr < greatest - least || current_p + descr <= least || current_p >= greatest
Right?

**Comment by GaborDrescher** [comment]
I would add brackets, but yes this should be all right, nice idea.

**Comment by GaborDrescher** [comment]
I created a pull request based on your proposal, #127

**Comment by ivmai** [comment]
Merged to master - da2fcda (I will copy it to release-7_2 and release-7_4 branches later)
